src/adapter/src/coord.rs

```diff
@@ -3522,6 +3540,55 @@ impl Coordinator {
             }
         }
     }
+
+    /// Prune all storage usage events from the [`MZ_STORAGE_USAGE_BY_SHARD`] table that are older
+    /// than `retention_period`.
+    ///
+    /// This method will read the entire contents of [`MZ_STORAGE_USAGE_BY_SHARD`] into memory
+    /// which can be expensive.
+    ///
+    /// DO NOT call this method outside of startup. The safety of reading at the current oracle read
+    /// timestamp and then writing at whatever the current write timestamp is (instead of
+    /// `read_ts + 1`) relies on the fact that there are no outstanding writes during startup.
+    ///
+    /// In practice, this method is always safe to call because group commit has builtin fencing,
+    /// and we never commit retractions to [`MZ_STORAGE_USAGE_BY_SHARD`] outside of this method. So
+    /// we don't have to worry about double retractions. Still, better safe than sorry.
+    async fn prune_storage_usage_events_on_startup(&mut self, retention_period: Duration) {
+        let id = self
+            .catalog()
+            .resolve_builtin_table(&MZ_STORAGE_USAGE_BY_SHARD);
+        let read_ts = self.get_local_read_ts().await;
+        let mut current_contents = self
+            .controller
+            .storage
+            .snapshot(id, read_ts.clone())
+            .await
+            .unwrap_or_terminate("cannot fail to fetch snapshot");
+        differential_dataflow::consolidation::consolidate(&mut current_contents);
+
+        let cutoff_ts = u128::from(read_ts).saturating_sub(retention_period.as_millis());
+        let mut expired = Vec::new();
+        for (row, diff) in current_contents {
+            assert_eq!(
+                diff, 1,
+                "consolidated contents should not contain retractions: ({row:#?}, {diff:#?})"
+            );
+            // This logic relies on the definition of `mz_storage_usage_by_shard` not changing.
+            let collection_timestamp = row.unpack()[3].unwrap_timestamptz();
```

Comment:
nit: maybe something like `row.unpack().get(3).expect("definition of mz_storage_by_shard changed")` to improve debugability?

Reply:
Done